workspace/src/main/kotlin/com/cosmotech/workspace/azure/WorkspaceServiceImpl.kt

```diff
@@ -43,94 +37,101 @@ import org.springframework.stereotype.Service
 @Suppress("TooManyFunctions")
 internal class WorkspaceServiceImpl(
     private val resourceLoader: ResourceLoader,
-    private val userService: UserApiService,
-    private val organizationService: OrganizationApiService,
     private val solutionService: SolutionApiService,
     private val azureStorageBlobServiceClient: BlobServiceClient,
     private val azureStorageBlobBatchClient: BlobBatchClient,
 ) : AbstractCosmosBackedService(), WorkspaceApiService {
 
-  private fun fetchUsers(userIds: Collection<String>): Map<String, User> =
-      userIds.toSet().map { userService.findUserById(it) }.associateBy { it.id!! }
+  private fun validateUser(workspace: Workspace) {
+    // Owner is always a valid user
+    if (workspace.ownerId == getCurrentAuthenticatedUserName()) {
+      logger.debug(
+          "Owner {} is authorized on Workspace {} - {}",
+          workspace.ownerId,
+          workspace.name,
+          workspace.id)
+      return
+    }
+    // User mail must be the same as upn attribute in token
+    val currentUserUPN = getCurrentAuthenticatedUserUPN()
+    logger.debug("Validating user with UPN: {}", currentUserUPN)
+    val workspaceUsers = workspace.users
+    if (workspaceUsers != null && !workspaceUsers.contains(currentUserUPN))
+        throw CsmAccessForbiddenException("You are not allowed to access this workspace")
+    if (workspaceUsers == null) {
+      logger.warn("No users list set on Workspace: {} - {}", workspace.name, workspace.id)
+    } else {
+      logger.debug(
+          "User {} is authorized on Workspace {} - {}",
+          currentUserUPN,
+          workspace.name,
+          workspace.id)
+    }
+  }
 
   override fun findAllWorkspaces(organizationId: String) =
```

Comment:
Maybe we should consider making `findAllWorkspaces(String)` return Workspaces with `null` users, owned by or containing the authenticated user ?

Reply:
+ [] + role Platform.Admin

Reply:
dones